src/shared/components/RichTable.tsx

```diff
@@ -0,0 +1,666 @@
+        const cell = accessor
+          ? typeof accessor === "function"
+            ? accessor(row)
+            : row[accessor]
+          : row[key as keyof D];
+        let props: PropsBag = { key };
+
+        return {
+          useCellProps() {
+            return props;
+          },
+          render() {
+            return render ? render(cell, row) : cell;
+          }
+        };
+      });
+
+      if (clickActions.length) {
+        props.onClick = (e: React.MouseEvent) => {
+          if (
+            !(e.target as Element)
+              .closest("td")
+              ?.classList.contains("RichTable_selectCheckbox")
+          ) {
+            clickActions.forEach(action => {
+              internalExecuteAction(action.name, [row]);
+            });
+          }
+        };
+      }
+
+      const selectRow = () => {
+        setSelected(prevSelected => {
+          prevSelected[i] = !prevSelected[i];
+          return prevSelected.slice();
+        });
+      };
+
+      if (selectable) {
+        cells.unshift({
+          useCellProps() {
+            return {
+              key: "select-row",
+              className: "RichTable_selectCheckbox"
+            };
+          },
+          render() {
+            return (
+              <FormCheck
+                custom
+                label={""}
+                type="checkbox"
+                aria-label="select row"
+                id={`RichTable-${id.current}-select-row-${i}`}
+                checked={selected[i] || false}
+                onChange={selectRow}
+              />
+            );
+          }
+        });
+      }
+
+      return {
+        useRowProps() {
+          return props;
+        },
+        data: row,
+        cells,
+        isSelected: selected[i],
+        setSelected: selectRow
+      };
+    });
+  }, [
+    columns,
+    data,
+    selected,
+    selectable,
+    clickActions,
+    internalExecuteAction
+  ]);
+
+  const makeRequest = (method: Method, row?: D, requestData?: any) => {
+    let requestUrl = url;
+    if (row !== null && row !== undefined) {
+      requestUrl = `${url}/${row[pk]}`;
+    }
+
+    return axios(requestUrl, {
+      method,
+      data: requestData
+    });
+  };
+
+  const bag: RichTableBag<D> = {
+    error,
+    header,
+    rows,
+    selected: filteredSelected,
+    page,
+    numPages,
+    setPage,
+    totalCount,
+    searchQuery,
+    setSearchQuery,
+    executeAction,
+    makeRequest
+  };
+
+  return bag;
+};
+
+/**
+ * The RichTable component, on the other hand, is a ready-to-use table with pagination, sorting, and search natively
+ * supported.
+ */
+const RichTable = <D extends object = {}>(config: RichTableProps<D>) => {
+  const {
+    header,
+    rows,
+    selected,
+    page,
+    setPage,
+    numPages,
+    totalCount,
+    setSearchQuery,
+    executeAction,
+  } = useRichTable(config);
+  const debouncer = useRef<number>(-1);
```

Comment:
We do have a debouncer hook in the shared folder. Any reason why that isn't being used here?

Reply:
The behaviour I was going for was an instant refresh when you clear the search query, which isn't possible with the debouncer hook.

Reply:
We can probably modify it to add that functionality. Something like:
```
export const useDebouncedCallback = <A extends any[]>(
  callback: (...args: A) => void | Promise<void>,
  wait: number
) => {
  // track args & timeout handle between calls
  const argsRef = useRef<A>();
  const timeout = useRef<ReturnType<typeof setTimeout>>();

  const cleanup = () => {
    if (timeout.current) {
      clearTimeout(timeout.current);
    }
  };

  // make sure our timeout gets cleared if
  // our consuming component gets unmounted
  useEffect(() => cleanup, []);

  return [
    useCallback(
      (...args: A) => {
        // capture latest args
        argsRef.current = args;

        // clear debounce timer
        cleanup();

        // start waiting again
        timeout.current = setTimeout(() => {
          if (argsRef.current) {
            callback(...argsRef.current);
          }
        }, wait);
      },
      [callback, wait]
    ),
    // add function to immediately trigger the callback
    useCallback(
      (...args: A) => {
        callback(...args);
        cleanup();
      },
      [callback]
    ),
  ];
};
```

Having a named hook instead of a bare ref should be more clear.

Reply:
7e0f3c1